### DISCUSSION TITLE
Feature Request: Feature Flags

### DISCUSSION THREAD

**Discussion by rhlabsuk** [supabase/supabase#27078] [Feature Requests]
Hi everyone, I've been creating an app with Supabase and really appreciate how easy things like authentication and data retrieval are. One thing I am looking to implement is Feature Flags to allow certain features of the app to be turned on/off if an issue is detected without going through a full release cycle, as well as A/B testing.

Would feature flags functionality be something that would be suitable for Supabase? it would be great to be able to show/hide functionality in our site via toggles in the Supabase dashboard as opposed to having something extra like Flagsmith to manage the flags.

While this could technically be done with queries before the page loads, it does comes with some pitfalls. Something native in Supabase would be great to just toggle features on/off quickly.

Would anyone else be interested to see something like this built into Supabase? Realtime seems like it would be a nice complement to this, too.

Thanks,
Ryan

**Comment by johhansantana** [comment]
would love this as well +1

**Comment by rnambs** [comment]
very interested in seeing this play out +1. if anyone has clean workarounds for this at the time being would love to learn more

**Comment by rhlabsuk** [comment]
Has anyone got a good way to implement this with the current feature set? it would be great to lock down certain parts of an app unless they are enabled, or could someone from Supabase see if this can be added as a feature?

**Comment by Daspy11** [comment]
Came from Google, +1

**Comment by elyobo** [comment]
> Something native in Supabase would be great to just toggle features on/off quickly.

This would still ultimately require "queries before the page loads", as that's how you'd get the data from Supabase, so it seems like this is more of something that you could build yourself.

**Comment by mrdjohnson** [comment]
For supabase-js users that need something quick; I just wrote a FeatureFlag class that uses supabase and [@carbon/feature-flags](https://www.npmjs.com/package/@carbon/feature-flags)

my solution requires top level await to be enabled on your project, modify this to display a loading screen or something if you have a setup where the call takes some time or you cannot enable top level await

DB Table:

```mermaid 
erDiagram
    FEATURES {
        string id "Primay key"
        boolean is_dev_enabled "NOT NULL"
        boolean is_enabled "NOT NULL"
    }
```

Code:

```typescript
import { enabled, merge } from '@carbon/feature-flags'

import { supabase } from '~/core/supabase' // this is just the supabase client

const { data: featureRows } = await supabase.from('Features').select('*')

const features = (featureRows || []).reduce(
  function (result, value) {
    result[value.id] = import.meta.env.DEV ? value.devEnabled : value.enabled
    return result
  },
  {} as Record<string, boolean>,
)

merge(features)

export class Features {
  static isEnabled = (key: string) => enabled(key)
}

```

example usage: `Features.isEnabled('feature-flag-key')` somewhere in the project

or just stick with carbon, remove the exported class, and do `enabled('feature-flag-key')`


note: results with SSR may vary
